pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -1062,17 +1069,26 @@ private void makeConsumerWrapper(String reason) {
   // This should be done during commit? We may not always commit when we build a segment....
   // TODO Call this method when we are loading the segment, which we do from table datamanager afaik
   private void updateCurrentDocumentCountMetrics() {
-    int currentRawDocs = _realtimeSegment.getNumDocsIndexed();
-    _serverMetrics.addValueToTableGauge(_tableName, ServerGauge.DOCUMENT_COUNT, (currentRawDocs - _lastUpdatedRawDocuments
-        .get()));
-    _lastUpdatedRawDocuments.set(currentRawDocs);
+
+    // When updating of metrics is enabled, numRowsIndexed can be <= numRowsConsumed. This is because when in this
+    // case when a new row with existing dimension combination comes in, we find the existing row and update metrics.
+
+    // Number of rows indexed should be used for DOCUMENT_COUNT metric, and also for segment flush. Whereas,
+    // Number of rows consumed should be used for consumption metric.
+    long rowsIndexed = _numRowsIndexed - _lastUpdatedRawDocuments.get();
+    _serverMetrics.addValueToTableGauge(_tableName, ServerGauge.DOCUMENT_COUNT,
+        rowsIndexed);
+    _lastUpdatedRawDocuments.set(_numRowsIndexed);
     final long now = now();
     final int rowsConsumed = _numRowsConsumed - _lastConsumedCount;
     final long prevTime = _lastConsumedCount == 0 ? _consumeStartTime : _lastLogTime;
     // Log every minute or 100k events
-    if (now - prevTime > TimeUnit.MINUTES.toMillis(TIME_THRESHOLD_FOR_LOG_MINUTES) || rowsConsumed >= MSG_COUNT_THRESHOLD_FOR_LOG) {
-      segmentLogger.info("Consumed {} events from (rate:{}/s), currentOffset={}, numRowsSoFar={}", rowsConsumed,
-            (float) (rowsConsumed) * 1000 / (now - prevTime), _currentOffset, _numRowsConsumed);
+    if (now - prevTime > TimeUnit.MINUTES.toMillis(TIME_THRESHOLD_FOR_LOG_MINUTES)
+        || rowsConsumed >= MSG_COUNT_THRESHOLD_FOR_LOG) {
+      segmentLogger.info(
+          "Consumed {} events from (rate:{}/s), currentOffset={}, numRowsConsumedSoFar={}, numRowsIndexedSoFar={}",
+          rowsConsumed, (float) (rowsConsumed) * 1000 / (now - prevTime), _currentOffset, _numRowsConsumed,
+          rowsIndexed);
```

Comment:
s/rowsIndexed/_numRowsIndexed/ ? As I understand it, rowsIndexed is the number of rows indexed since the last update, but _numRowsIndexed is the number of rows indexed since start of segment.

Reply:
Renamed (will revisit the atomicLong vs volatile later).